BreadPlayer.Views.UWP/ViewModels/ShellViewModel.cs

```diff
@@ -350,40 +353,41 @@ private void SetNowPlayingSong()
         {
-            var playingCollection = GetPlayingCollection();
-            NowPlayingQueue = playingCollection;
+            var playingCollection = Playlist;
+            NowPlayingQueue = Playlist;
+            Mediafile toPlayFile = null;
             if (playingCollection != null && playingCollection.Any())
             {
                 try
                 {
-                    int indexOfCurrentlyPlayingFile = -1;
-                    if (playingCollection.Any(t => t.State == PlayerState.Playing))
+                    if (IsSourceGrouped)
                     {
-                        indexOfCurrentlyPlayingFile = playingCollection.IndexOf(playingCollection.FirstOrDefault(t => t.State == PlayerState.Playing));
+                        toPlayFile = GetNextSongInGroup();
                     }
-
-                    Mediafile toPlayFile = null;
-                    if (Shuffle)
+                    else
                     {
-                        if (_shuffledList.Count < playingCollection.Count 
-                            || _shuffledList == null
-                            || indexOfCurrentlyPlayingFile > playingCollection.Count - 2)
+                        int indexOfCurrentlyPlayingFile;
+                        if (playingCollection.Any(t => t.State == PlayerState.Playing))
+                        {
+                            indexOfCurrentlyPlayingFile =
+                                playingCollection.IndexOf(
+                                    playingCollection.FirstOrDefault(t => t.State == PlayerState.Playing));
+                        }
+                        else
                         {
-                            _shuffledList = await ShuffledCollection();
-                            
                             indexOfCurrentlyPlayingFile = 0;
                         }
-                        toPlayFile = _shuffledList?.ElementAt(indexOfCurrentlyPlayingFile + 1);
-                    }
-                    else if (IsSourceGrouped)
-                    {
-                        toPlayFile = GetNextSongInGroup();
-                    }
-                    else
-                    {
-                        toPlayFile = indexOfCurrentlyPlayingFile <= playingCollection.Count - 2 && indexOfCurrentlyPlayingFile != -1 ? playingCollection.ElementAt(indexOfCurrentlyPlayingFile + 1) : Repeat == "Repeat List" || isNext ? playingCollection.ElementAt(0) : null;
+                        var indexToPlay = indexOfCurrentlyPlayingFile;
+                        do
+                        {
+                            indexToPlay += 1;
+                            if (indexToPlay > playingCollection.Count - 1)
+                            {
+                                indexToPlay = 0;
+                            }
+                            toPlayFile = playingCollection[indexToPlay];
+                        } while (Shuffle && toPlayFile.SkipOnShuffle);
```

Comment:
The do-while loop here will get the next index, and then check if that song has the skip on shuffle property. If it does, it will get the song after that. It isn't perfect - if your entire playlist is (for *whatever* reason) set to skip on shuffle, you'll have issues. I haven't worked that one out yet.